### PR TITLE
Undo file actions

### PR DESCRIPTION
**Pull request by iamsethsamuel**
Last actions on files such as moving, rename, creating and deleting a file can now be undone with Ctrl+Z

### CODE REVIEW

File: apps/remix-ide/src/app/files/fileManager.ts [hunk 1]
```diff
@@ -861,7 +885,7 @@ class FileManager extends Plugin {
    * @returns {void}
    */
   
-  async moveDir(src: string, dest: string) {
+   async moveDir(src: string, dest: string) {
```
Comment (LianaHus) [inline apps/remix-ide/src/app/files/fileManager.ts:888]: no need for space here

File: apps/remix-ide/src/app/files/fileManager.ts [hunk 2]
```diff
@@ -877,11 +901,93 @@ class FileManager extends Plugin {
...
+  /**
+   * Reverts the last action that was performed by the user
+   * @returns {void}
+  */
+
+  async revertFileAction(redo: boolean){
+    this.isFromLastAction = true
+
+    let lastAction = this.actions[this.actions.length - 1]
+    if(redo){
```
Comment (LianaHus) [inline apps/remix-ide/src/app/files/fileManager.ts:935]: space needed after if. also below

File: apps/remix-ide/src/app/files/fileManager.ts (hunk 1 above)
Comment (iamsethsamuel) [reply apps/remix-ide/src/app/files/fileManager.ts:888]: I have removed it

File: apps/remix-ide/src/app/files/fileManager.ts (hunk 2 above)
Comment (iamsethsamuel) [reply apps/remix-ide/src/app/files/fileManager.ts:935]: I have added it

File: apps/remix-ide/src/app/files/fileManager.ts [hunk 3]
```diff
@@ -396,14 +411,21 @@ class FileManager extends Plugin {
       path = this.limitPluginScope(path)
       await this._handleExists(path, `Cannot remove file or directory ${path}`)
       const provider = this.fileProviderOf(path)
+
+      if(await this.isDirectory(path)){
```
[the remark this answers is not in the document]
Comment (iamsethsamuel) [reply apps/remix-ide/src/app/files/fileManager.ts:415]: I have done that

**Comment by iamsethsamuel** [comment]
I have fixed the merge conflict

**Comment by iamsethsamuel** [comment]
Please can someone merge this so that I can send the next PR?

**Comment by bunsenstraat** [comment]
> Please can someone merge this so that I can send the next PR?

Yes, but we need to discuss some things about it.

**Comment by bunsenstraat** [comment]
Two things
- when I switch workspace it should remove the recorded actions otherwise you'd be undoing it in another workspace
- the undo action should be available globally: so ie you are doing a file rename, and then click somewhere else it should still undo, even when the cursor is in the editor. so these things should be linked if we can do that... might be hard. or we have another way except ctrl-z. Macs btw don't use CTRL. 
- the undo actions should have a modal ( I think ) telling you you're undoing something.

**Comment by iamsethsamuel** [comment]
> 

Okay, I modelled it after VSCode's implementation. Which I think most people are used to so if the behaviour is different here I guess it will create a confusion. I think the file explorer in windows and Linux also require focus to work. So imagine I working on the editor and want to undo my code and I undo a rename. That will be weird.

As for the dialog, you can always redo so I don't see the need for it.

The workspace is very important, let me see if I can implement it.

And I don't have a Mac so I don't know how things work there. If someone can help implement it that will be awesome.

Just my thoughts, but we can do it your way

**Comment by bunsenstraat** [comment]
I think it would be good to have a call with the team and discuss how we could do it. Would that fit for you?

**Comment by bunsenstraat** [comment]
Thank you for making this effort btw!! It's a great idea.

**Comment by iamsethsamuel** [comment]
> I think it would be good to have a call with the team and discuss how we could do it. Would that fit for you?

Yes, when will you be free for the call

**Comment by iamsethsamuel** [comment]
Please can someone review this?

**Review by ioedeveloper — APPROVED** [review]

**Pull request closed without merging**